### ISSUE TITLE
自己数据集上作测试效果差

### ISSUE THREAD

**Issue by kuangcaomvp** [ZrrSkywalker/MonoDETR#27]
![image](https://user-images.githubusercontent.com/26867020/216211461-d9435962-49bf-4db3-aef9-35931d1e4f6c.png)
作者你好，我在kitti上做的训练，然后在深云相机上作的测试，发现近景目标偏移很严重，这个是什么原因导致的。
相机内参使用的ros内参标定程序标定的，图片分辨率1920*1080。

**Comment by zzylovexx** [comment]
請問你是修改那些地方讓他在你的測試集上跑呢

**Comment by kuangcaomvp** [comment]
> 請問你是修改那些地方讓他在你的測試集上跑呢

![11](https://user-images.githubusercontent.com/26867020/219597340-0bfab78b-4429-47ff-80f6-16e54a12514a.png)

您好，在kitti数据集上测试效果很好，我用的是深云相机，我就修改了内参为自己的内参数据，具体的测试代码如下，我测试发现近景的3d框偏移大概一个车身，远景的看着还可以，不知道什么原因，能不能麻烦帮忙看看是不是代码哪个地方写错了还是其他什么原因，谢谢


import warnings

from lib.datasets.kitti.kitti_utils import get_affine_transform
from lib.helpers.decode_helper import extract_dets_from_outputs, get_heading_angle
from utils.vis3d import vis_3d_with_calib_smoke

warnings.filterwarnings("ignore")

import os
import sys
import torch

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
ROOT_DIR = os.path.dirname(BASE_DIR)
sys.path.append(ROOT_DIR)

import yaml
import argparse

from lib.helpers.model_helper import build_model
from lib.helpers.utils_helper import set_random_seed
import rospy
import numpy as np
from sensor_msgs.msg import Image as Img
import cv2
from PIL import Image
import time

parser = argparse.ArgumentParser(description='Depth-aware Transformer for Monocular 3D Object Detection')
parser.add_argument('--config', default='configs/monodetr.yaml', help='settings of detection in yaml format')
parser.add_argument('-e', '--evaluate_only', action='store_true', default=False, help='evaluation only')

args = parser.parse_args()
def img_to_rect(u, v, depth_rect, calib):
    """
    :param u: (N)
    :param v: (N)
    :param depth_rect: (N)
    :return:
    """
    a = np.array([u * depth_rect, v * depth_rect, depth_rect], dtype=np.float32).reshape((3, 1))
    k_inv = np.linalg.inv(calib[:3, :3])
    pts_rect = np.dot(k_inv, a).T

    # a = np.array([u, v, 1], dtype=np.float32).reshape((3, 1))
    # a = np.matmul(target, a)
    # a = np.concatenate([a, np.array([[1]])], dtype=np.float32)
    # a = a * depth_rect
    # k_inv = np.linalg.inv(calib[:3, :3])
    # pts_rect = np.dot(k_inv, a).T
    return pts_rect


def alpha2ry(alpha, u, calib):
    """
    Get rotation_y by alpha + theta - 180
    alpha : Observation angle of object, ranging [-pi..pi]
    x : Object center x to the camera center (x-W/2), in pixels
    rotation_y : Rotation ry around Y-axis in camera coordinates [-pi..pi]
    """
    cu = calib[0, 2]
    fu = calib[0, 0]
    ry = alpha + np.arctan2(u - cu, fu)

    if ry > np.pi:
        ry -= 2 * np.pi
    if ry < -np.pi:
        ry += 2 * np.pi

    return ry


def decode_detections(dets, img_size, calibs, cls_mean_size, threshold):
    '''
    NOTE: THIS IS A NUMPY FUNCTION
    input: dets, numpy array, shape in [batch x max_dets x dim]
    input: img_info, dict, necessary information of input images
    input: calibs, corresponding calibs for the input batch
    output:
    '''

    preds = []
    for j in range(dets.shape[0]):  # max_dets
        cls_id = int(dets[j, 0])
        score = dets[j, 1]
        if score < threshold:
            continue

        # 2d bboxs decoding
        x = dets[j, 2] * img_size[0]
        y = dets[j, 3] * img_size[1]
        w = dets[j, 4] * img_size[0]
        h = dets[j, 5] * img_size[1]
        bbox = [x - w / 2, y - h / 2, x + w / 2, y + h / 2]

        # 3d bboxs decoding
        # depth decoding
        depth = dets[j, 6]

        # dimensions decoding
        dimensions = dets[j, 31:34]
        dimensions += cls_mean_size[int(cls_id)]

        # positions decoding
        x3d = dets[j, 34] * img_size[0]
        y3d = dets[j, 35] * img_size[1]
        locations = img_to_rect(x3d, y3d, depth, calibs).reshape(-1)
        locations[1] += dimensions[0] / 2

        # heading angle decoding
        alpha = get_heading_angle(dets[j, 7:31])
        ry = alpha2ry(alpha, x, calibs)

        score = score * dets[j, -1]
        preds.append([cls_id, alpha] + bbox + dimensions.tolist() + locations.tolist() + [ry, score])
    return preds


class Manager:
    def __init__(self, cfg, model, device):
        self.sub = rospy.Subscriber("/tztek/camera0", Img, self.callback, queue_size=10)

        self.cfg = cfg
        self.model = model
        self.device = device
        self.max_objs = 50
        self.num_classes = 4
        self.cls_mean_size = np.zeros((self.num_classes, 3), dtype=np.float32)

        ### calib_cam_to_cam.txt 中的 P_rect_02
        self.resolution = np.array([1280, 384])  # W * H
        self.mean = np.array([0.485, 0.456, 0.406], dtype=np.float32)
        self.std = np.array([0.229, 0.224, 0.225], dtype=np.float32)
        self.calib = np.array(
            [1.35471429e+03, 0.00000000e+00, 1.00673454e+03, 0,
             0.00000000e+00, 1.36207264e+03, 5.13843423e+02, 0,
             0.00000000e+00, 0.00000000e+00, 1.00000000e+00, 0], dtype=np.float32).reshape([3, 4])
        self.K = self.calib[:3, :3]
        self.pub_image = rospy.Publisher('/image_with_box', Img, queue_size=10)

        # self.distCoeff = np.array(
        #     [-0.45963538, 0.30819234, 0.00244484, 0.00836784],
        #     dtype=np.float32).reshape([1, 4])
        # self.ncm, self.roi = cv2.getOptimalNewCameraMatrix(self.K, self.distCoeff, (1920, 1080), 1, (1920, 1080))
        # self.mapx, self.mapy = cv2.initUndistortRectifyMap(self.K, self.distCoeff, None, self.ncm, (1920, 1080), 5)
        # self.calib[:3, :3] = self.ncm

    def image_ros(self, image, frameId):
        image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
        ros_image = Img()
        ros_image.header.stamp = rospy.Time.now()
        ros_image.header.frame_id = frameId
        ros_image.encoding = 'rgb8'
        ros_image.height = image.shape[0]
        ros_image.width = image.shape[1]
        ros_image.step = image.shape[1] * image.shape[2]
        ros_image.data = np.array(image).tobytes()
        return ros_image

    def callback(self, msg):
        t1 = time.time()
        img = Image.frombuffer(mode='RGB', size=(msg.width, msg.height), data=msg.data)
        img_c = np.array(img)
        """
        img_c = cv2.undistort(img_c, self.K, self.distCoeff, None, self.ncm)
        # img_c = cv2.remap(img_c, self.mapx, self.mapy, cv2.INTER_LINEAR)
        x, y, w, h = self.roi
        w = int(w // 10 * 10)
        h = int(h // 10 * 10)
        img_c = img_c[y:y + h, x:x + w]
        img = Image.fromarray(img_c, mode='RGB')
        """
        img_size = np.array(img.size)

        center = np.array(img_size) / 2
        crop_size = img_size
        trans, trans_inv = get_affine_transform(center, crop_size, 0, self.resolution, inv=1)
        img = img.transform(tuple(self.resolution.tolist()),
                            method=Image.AFFINE,
                            data=tuple(trans_inv.reshape(-1).tolist()),
                            resample=Image.BILINEAR)
        img = np.array(img).astype(np.float32) / 255.0
        img = (img - self.mean) / self.std
        img = img.transpose(2, 0, 1)  # C * H * W

        inputs = torch.Tensor(img).to(self.device)
        inputs = torch.unsqueeze(inputs, dim=0)
        calibs = torch.Tensor(self.calib).to(self.device)
        calibs = torch.unsqueeze(calibs, dim=0)
        img_sizes = torch.Tensor(img_size).to(self.device)
        img_sizes = torch.unsqueeze(img_sizes, dim=0)
        t2 = time.time()
        print('image process time:' + str(t2 - t1))
        with torch.no_grad():
            outputs = self.model(inputs, calibs, img_sizes)
            t3 = time.time()
            print('model process time:' + str(t3 - t2))
            dets = extract_dets_from_outputs(outputs=outputs, K=self.max_objs, topk=self.max_objs)
            dets = dets.detach().cpu().numpy()[0]
            dets = decode_detections(
                dets=dets,
                img_size=img_size,
                calibs=self.calib,
                cls_mean_size=self.cls_mean_size,
                threshold=self.cfg.get('threshold', 0.2))

        #     obj_list = [KittiInfo(i) for i in dets]
        #     gt_boxes3d_corner = []
        #     for obj in obj_list:
        #         center = np.array([obj.tx, obj.tz, obj.ty])
        #         lwh = np.array([obj.l, obj.w, obj.h])
        #         axis_angles = np.array([0, 0, obj.rot_global + 1e-10])
        #         rot = open3d.geometry.get_rotation_matrix_from_axis_angle(axis_angles)
        #         box3d = open3d.geometry.OrientedBoundingBox(center, rot, lwh)
        #         corners_3D = np.asarray(box3d.get_box_points())
        #         gt_boxes3d_corner.append(corners_3D)
        # gt_boxes3d_corner = np.array(gt_boxes3d_corner).reshape(-1, 8, 3)
        # #
        # hight_with_box = draw_polygons(self.bev, gt_boxes3d_corner[:, [3, 6, 4, 5], :2])
        image_with_box = vis_3d_with_calib_smoke(img_c, self.calib, dets)

        # cv2.imshow('1', image_with_box)
        # cv2.waitKey(1)
        # concat_height_image = np.zeros(shape=(max(W, 375), H + 1242, 3), dtype=np.uint8)
        # concat_height_image[:375, :1242, :] = image_with_box
        # concat_height_image[:W, 1242:, :] = hight_with_box
        #
        # # ros_hight_with_box = self.image_ros(hight_with_box, msg.header.frame_id)
        ros_image_with_box = self.image_ros(image_with_box, msg.header.frame_id)
        t4 = time.time()
        print('back process time:' + str(t4 - t3))
        # ros_concat_height_image = self.image_ros(concat_height_image, msg.header.frame_id)
        # # self.pub_height.publish(ros_hight_with_box)
        self.pub_image.publish(ros_image_with_box)
        # self.pub_concate.publish(ros_concat_height_image)
        # print('all:', str(time.time() - b))


if __name__ == '__main__':
    cfg = yaml.load(open(args.config, 'r'), Loader=yaml.Loader)
    set_random_seed(cfg.get('random_seed', 444))
    # build model
    model, _ = build_model(cfg['model'])
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    checkpoint_path = 'runs/monodetr/checkpoint_best.pth'
    checkpoint = torch.load(checkpoint_path, device)
    model.load_state_dict(checkpoint['model_state'])
    model = model.to(device)
    torch.set_grad_enabled(False)
    model.eval()

    rospy.init_node('map', anonymous=True)
    Manager(cfg, model, device)
    rospy.spin()
